provider/server/exec.go

```diff
@@ -5,11 +5,150 @@ package server
+	"github.com/moby/term"
+	"k8s.io/client-go/tools/remotecommand"
 )
 
-func (s *Server) Exec(ctx context.Context, req *ori.ExecRequest) (*ori.ExecResponse, error) {
+const (
+	StreamCreationTimeout = 30 * time.Second
+	StreamIdleTimeout     = 2 * time.Minute
+)
+
+type executorExec struct {
+	Libvirt         *libvirt.Libvirt
+	ExecRequest     *iri.ExecRequest
+	VirshExecutable string
+}
+
+func (s *Server) Exec(ctx context.Context, req *iri.ExecRequest) (*iri.ExecResponse, error) {
+	machineID := req.MachineId
+	log := s.loggerFrom(ctx, "MachineID", machineID)
+
+	log.V(1).Info("Inserting request into cache")
+	token, err := s.execRequestCache.Insert(req)
+	if err != nil {
+		return nil, err
+	}
+
+	log.V(1).Info("Returning url with token")
+	return &iri.ExecResponse{
+		Url: s.buildURL("exec", token),
+	}, nil
+}
+
+func (s *Server) ServeExec(w http.ResponseWriter, req *http.Request, token string) {
+	ctx := req.Context()
+	log := logr.FromContextOrDiscard(ctx)
+
+	request, ok := s.execRequestCache.Consume(token)
+	if !ok {
+		log.V(1).Info("Rejecting unknown / expired token")
+		http.NotFound(w, req)
+		return
+	}
+
+	exec := executorExec{
+		Libvirt:         s.libvirt,
+		ExecRequest:     request,
+		VirshExecutable: s.virshExecutable,
+	}
+
+	handler, err := remotecommandserver.NewExecHandler(exec, remotecommandserver.ExecHandlerOptions{
+		StreamCreationTimeout: StreamCreationTimeout,
+		StreamIdleTimeout:     StreamIdleTimeout,
+	})
+	if err != nil {
+		log.Error(err, "Error creating exec handler")
+		code := http.StatusInternalServerError
+		http.Error(w, http.StatusText(code), code)
+		return
+	}
+
+	handler.Handle(w, req, remotecommandserver.ExecOptions{})
+}
+
+func (e executorExec) Exec(ctx context.Context, in io.Reader, out io.WriteCloser, resize remotecommand.TerminalSizeQueue) error {
+	var wg sync.WaitGroup
+
+	machineID := e.ExecRequest.MachineId
+	log := logr.FromContextOrDiscard(ctx).WithName(machineID)
+
+	_, err := e.Libvirt.DomainLookupByUUID(libvirtutils.UUIDStringToBytes(machineID))
+	if err != nil {
+		if !libvirtutils.IsErrorCode(err, libvirt.ErrNoDomain) {
+			return fmt.Errorf("error looking up domain: %w", err)
+		}
+		return fmt.Errorf("machine %s has not yet been synced", machineID)
+	}
+
+	uri, err := e.Libvirt.ConnectGetUri()
+	if err != nil {
+		return fmt.Errorf("error getting connection uri")
+	}
+
+	cmd := exec.CommandContext(ctx, e.VirshExecutable, "-c", strings.TrimSpace(uri), "console", machineID)
+
+	f, err := pty.Start(cmd)
+	if err != nil {
+		return fmt.Errorf("error starting command: %w", err)
+	}
+
+	// Wrap the input stream with an escape proxy. Escape Sequence Ctrl + ] = 29
+	inputReader := term.NewEscapeProxy(in, []byte{29})
+
+	wg.Add(2)
+	// ReadInput: go routine to read the input from the reader, and write to the terminal.
+	go func() {
+		defer wg.Done()
+
+		buf := make([]byte, 1024)
+		for {
+			n, err := inputReader.Read(buf)
+			if err != nil {
+				if _, ok := err.(term.EscapeError); ok {
```

Comment:
it is old way. it is recommended use errors.Is currently.

Reply:
i am little confuse because error interface requires Error() method which returns string: https://pkg.go.dev/builtin#error

```golang
type error interface {
	Error() [string](https://pkg.go.dev/builtin#string)
}
```

Reply:
EscapError is structure type and it cannot be used with errors.Is and it require create object of specific type. We decide keep current solution because benefits isn't huge.